topostats/tracing/ordered_tracing.py

```diff
@@ -0,0 +1,729 @@
+        """
+        # node fwhms can be a list of different lengths so cannot use np arrays
+        under_idxs = []
+        over_idxs = []
+        for node_fwhms in fwhms:
+            order = np.argsort(node_fwhms)
+            under_idxs.append(order[0])
+            over_idxs.append(order[-1])
+        return under_idxs, over_idxs
+
+    def check_node_errorless(self) -> bool:
+        """
+        Check if an error has occurred while processing the node dictionary.
+
+        Returns
+        -------
+        bool
+            Whether the error is present.
+        """
+        for _, vals in self.nodestats_dict.items():
+            if vals["error"]:
+                return False
+        return True
+
+    def run_nodestats_tracing(self) -> tuple[list, dict, dict]:
+        """
+        Run the nodestats tracing pipeline.
+
+        Returns
+        -------
+        tuple[list, dict, dict]
+            A list of each molecules ordered trace coordinates, the ordered_traicing stats, and the images.
+        """
+        self.ordered_traces, self.images = self.compile_trace()
+        self.grain_tracing_stats["num_mols"] = len(self.ordered_traces)
+
+        ordered_trace_data = {}
+        for i, mol_trace in enumerate(self.ordered_traces):
+            self.mol_tracing_stats["circular"] = linear_or_circular(mol_trace)
+            ordered_trace_data[f"mol_{i}"] = {
+                "ordered_coords": mol_trace,
+                "heights": self.image[mol_trace[:, 0], mol_trace[:, 1]],
+                "distances": coord_dist(mol_trace[0]),
+                "mol_stats": self.mol_tracing_stats,
+            }
+
+        return ordered_trace_data, self.grain_tracing_stats, self.images
+
+
+class OrderedTraceTopostats:
+    """
+    Order single pixel thick skeleton coordinates via TopoStats.
+
+    Parameters
+    ----------
+    image : npt.NDArray
+        A cropped image array.
+    skeleton : npt.NDArray
+        The pruned skeleton mask array.
+    """
+
+    def __init__(
+        self,
+        image,
+        skeleton,
+    ) -> None:
+        """
+        Initialise the OrderedTraceTopostats class.
+
+        Parameters
+        ----------
+        image : npt.NDArray
+            A cropped image array.
+        skeleton : npt.NDArray
+            The pruned skeleton mask array.
+        """
+        self.image = image
+        self.skeleton = skeleton
+        self.grain_tracing_stats = {
+            "num_mols": 1,
+            "circular": None,
+        }
+        self.mol_tracing_stats = {"circular": None}
+
+        self.images = {
+            "ordered_traces": np.zeros_like(image),
+            "all_molecules": skeleton.copy(),
+            "over_under": skeleton.copy(),
+            "trace_segments": skeleton.copy(),
+        }
+
+    @staticmethod
+    def get_ordered_traces(disordered_trace_coords: npt.NDArray, mol_is_circular: bool) -> list:
```

Comment:
1. I think I just like to make things static if possible, let me know what is best practice if this isn't it.
2. the argwhere obtains the coordinates from the skeleton mask which is then used in later functions.
3. And it can be changed if that's best practice, but I'd like to know the rules to make this consistent across my changes. I also don't understand the last bit - is this one of those "because it doesn't change throughout the it should be an attribute" thing?